### PR TITLE
[server, dashboard] Fix (offline) headless log access for non-owners

### LINKED ISSUE #5569: [staging] Access to (offline) headless logs does not work for non-owners
With the introduction of "Teams & Projects" we use the API for accessing logs for headless workspaces which are offline the first time. This has been tested and basically works, but apparently not for non-owners.

### PR DESCRIPTION
**Pull request by geropl**
## Description
This contains two commits:
 - don't spam us with logs requests if something is obviously wrong: https://github.com/gitpod-io/gitpod/commit/54ea45d25c31607d3bbd256fc436ca89c9feca08
 - use proper userId when requesting prebuilds logs which have been triggered by a different user: https://github.com/gitpod-io/gitpod/commit/207bd07c1440f5a0d140e18dfcdf280c10a06142
image

## Related Issue(s)
Fixes #5569.

## How to test
1. create two accounts A and B, in different containers/sessions. Get "T & P" permission in both
1. A: create a new team and project, select a repo with prebuids configured, trigger a prebuild and wait until it finishes
1. A: after the prebuild succeeded, access the logs and see that it works: :heavy_check_mark: 
1. A: invite B into the test team
1. B: access invite link
1. B: find the prebuild and access it's log: :heavy_check_mark: 

## Release Notes
```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

Associated issue: *#5569*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/src/workspace/headless-log-service.ts [hunk 1]
```diff
@@ -124,14 +124,15 @@ export class HeadlessLogService {
...
      * @param taskId
      * @returns
      */
-    async getHeadlessLogDownloadUrl(userId: string, wsi: WorkspaceInstance, taskId: string): Promise<string | undefined> {
+    async getHeadlessLogDownloadUrl(userId: string, wsi: WorkspaceInstance, ownerId: string, taskId: string): Promise<string | undefined> {
         try {
             return await new Promise<string>((resolve, reject) => {
                 const req = new LogDownloadURLRequest();
-                req.setOwnerId(userId);
+                req.setOwnerId(ownerId);
```
Comment (AlexTugarev) [inline components/server/src/workspace/headless-log-service.ts:135]: 👏🏻  thank you so much!!!

File: components/dashboard/src/components/PrebuildLogs.tsx [hunk 2]
```diff
@@ -128,10 +128,16 @@ export function watchHeadlessLogs(instanceId: string, onLog: (chunk: string) =>
       return;
     }
 
-    const retry = async (reason: string, err?: Error) => {
+    const initialDelaySeconds = 1;
+    let delayInSeconds = initialDelaySeconds;
+    const retryBackoff = async (reason: string, err?: Error) => {
+      const backoffFactor = 1.2;
+      const maxBackoffSeconds = 5;
+      delayInSeconds = Math.min(delayInSeconds * backoffFactor, maxBackoffSeconds);
```
Comment (AlexTugarev) [inline components/dashboard/src/components/PrebuildLogs.tsx:136]: 👍🏻

File: components/dashboard/src/components/PrebuildLogs.tsx [hunk 3]
```diff
@@ -195,10 +201,32 @@ export function watchHeadlessLogs(instanceId: string, onLog: (chunk: string) =>
       }
     } catch(err) {
       reader?.cancel().catch(console.debug);
-      await retry("error while listening to stream", err);
+      if (err.code === 400) {
+        // sth is really off, and we _should not_ retry
+        console.error("stopped watching headless logs", err);
```
Comment (AlexTugarev) [inline components/dashboard/src/components/PrebuildLogs.tsx:206]: 👍🏻

File: components/server/src/workspace/headless-log-service.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/workspace/headless-log-service.ts:135]: :bow:  Sorry for this. Has been introduced with recent proxy endpoint I think.

**Comment by AlexTugarev** [comment]
Awesome! It works as advertised. 

Just one thing, in general it appears that short living prebuilds aren't reporting logs at all. If tasks do just a quick `yarn add typescript` for instance, we're not seeing any logs, and again no logs are stored.

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: cc121e83484a932efe40e1c6512a07f885aee1ab</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5570#issuecomment-914265754" title="LGTM">AlexTugarev</a>*

Associated issue: *#5569*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by geropl** [comment]
> Just one thing, in general it appears that short living prebuilds aren't reporting logs at all. If tasks do just a quick yarn add typescript for instance, we're not seeing any logs, and again no logs are stored.

Hm, that's strange - but I already have a suspicion. It was introduced with one of the "fixes" for the "stuck in stopping". :cry: 
Would you mind creating an issue with a reference, possibly?

**Pull request merged**